contracts/StorageManager.sol

```diff
@@ -172,177 +156,200 @@ contract StorageManager {
-                agreement.numberOfPeriodsDeposited = 0;
-                agreement.startDate = 0;
-                // check if Offer is still active
-                if(offerRegistry[msg.sender].availableCapacity != 0) {
+        for (uint8 i = 0; i < agreementReferences.length; i++) {
+        Agreement storage agreement = offer.agreementRegistry[agreementReferences[i]];
+            require(agreement.size != 0, "StorageManager: Agreement for this Offer doesn't exist");
+
+            // TODO: [Q] Should we disallow to payout from inactive agreement? Can it cause some deadlock?
+            require(agreement.lastPayoutDate != 0, "StorageManager: Agreement is inactive");
+
+            uint256 spentFunds = _calculateSpentFunds(agreement);
+            agreement.availableFunds = agreement.availableFunds.sub(spentFunds);
+            toTransfer = toTransfer.add(spentFunds);
+
+            // Agreement ran out of funds ==> Agreement is expiring
+            if (agreement.availableFunds < agreement.billingPrice * agreement.size) {
+                // Agreement becomes inactive
+                agreement.lastPayoutDate = 0;
+
+                // TODO: [Q] Should we automatically return any remaining agreement.availableFunds?
+                // Check if Offer is still active
+                if (offer.availableCapacity != 0) {
                     //ALLOW_OVERFLOW reasoning: see: REF_CAPACITY
                     // add back capacity
-                    offerRegistry[msg.sender].availableCapacity = offerRegistry[msg.sender].availableCapacity + agreement.size;
+                    offer.availableCapacity = offer.availableCapacity + agreement.size;
                 }
-            } else {
-                // NO_OVERFLOW reasoning: startDate is always less than now. agreement.billingPeriod is verified not to be 0 in function: newRequest
-                uint256 periodsPast = (now - agreement.startDate) / agreement.billingPeriod;
-                ////NO_OVERFLOW reasoning: numberOfPeriodsWithdrawn is always less than or equal to periodsPast and REF_MAX_TRANSFER
-                toTransfer = toTransfer.add((periodsPast - agreement.numberOfPeriodsWithdrawn) * agreement.billingPrice);
-                //SAFE_CAST & NO_OVERFLOW reasoning: periodsPast is always less than numberOfPeriodsDeposited (as agreement is not expired), and numberOfPeriodsDeposited is uint64
-                agreement.numberOfPeriodsWithdrawn += uint64(periodsPast);
+
+                emit AgreementStopped(agreementReferences[i]);
+            } else {// Provider called this during active agreement which has still funds to run
+                agreement.lastPayoutDate = uint128(_time());
             }
-            emit AgreementFundsPayout(agreementReferences[i]);
+
+            emit AgreementFundsPayout(agreementReferences[i], spentFunds);
         }
-        msg.sender.transfer(toTransfer);
+        provider.transfer(toTransfer);
```

Comment:
What about check if toTransfer > 0? It is only to caller spent less gas...
require (toTransfer > 0, "nothing to payout");

Suggestion to not use transfer because https://eips.ethereum.org/EIPS/eip-1884
(bool success, ) = provider.call.value(toTransfer)("");
require(success, "Transfer failed.");